### ISSUE TITLE
Info Content App - Links - Lots of 'This document is published but is not in the cache' messages one for every 'language' in non variant site

### ISSUE THREAD

**Issue by marcemarc** [umbraco/Umbraco-CMS#15361]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

12.3.3

### Bug summary

Well, this is a curious thing.

I have an Umbraco site with multiple Territories, set up as different sites within Umbraco, those Territories have different culture/languages, and so these are added to the Languages section of Umbraco and applied using the Cultures and Hostnames option for each root node.

The site doesn't use variants, it's not a 1-1 scenario, and there are pages on one territory that aren't in another territory - hence the separate branches.

But when I visit one of the pages on the site, switch to the Info content app, and have a look at the Link presented to the editors I see this:

![image](https://github.com/umbraco/Umbraco-CMS/assets/260802/802ef9f3-e612-4d78-8497-20f394ae2e2c)

This is confusing editors, as it seems like there is 'something wrong'.

People often mixup territories and cultures in their heads (let's not even mention territories with multiple languages - looking at you Switzerland!) - but although here the message is saying IF you requested the Url of this page on a page which had the culture of Canada, then although the underlying page is published, it's not available in the cache...

... but the editor thinks - how is this published in Canada? there is no Canadian Destinations page...

This kind of information would I think make sense to editors and be super useful too, if the site was using Variants, if there was 1-1, for languages rather than territories, then it would be useful to see all the urls in one place for a node and report which ones were published etc...

... but I'm not using Variants!

I thought this might be easily turned off with a custom UrlProvider - eg these 'other urls' were being added in via the 'Other Urls'  functionality of a Url Provider, but it turns out this is not the case.

Apologies if I am doing something stupid here!

## Digging a bit deeper into this:
The schenanigans are occurring in the GetContentUrlsByCultureAsync UrlProviderExtension:

https://github.com/umbraco/Umbraco-CMS/blob/14a64857df3edbd23558e7d741316db0ddccc098/src/Umbraco.Core/Routing/UrlProviderExtensions.cs#L107

This is running the UrlProvider multiple times, once for each culture:

![image](https://github.com/umbraco/Umbraco-CMS/assets/260802/40ced57e-c3e2-41c4-891a-4d40577d17ff)

It looks like this is 'by design'

```
   // build a list of URLs, for the back-office
        // which will contain
        // - the 'main' URLs, which is what .Url would return, for each culture
        // - the 'other' URLs we know (based upon domains, etc)
        //
        // need to work through each installed culture:
        // on invariant nodes, each culture returns the same URL segment but,
        // we don't know if the branch to this content is invariant, so we need to ask
        // for URLs for all cultures.
        // and, not only for those assigned to domains in the branch, because we want
        // to show what GetUrl() would return, for every culture.
        var urls = new HashSet<UrlInfo>();
        var cultures = localizationService.GetAllLanguages().Select(x => x.IsoCode).ToList();

        // get all URLs for all cultures
        // in a HashSet, so de-duplicates too
        foreach (UrlInfo cultureUrl in await GetContentUrlsByCultureAsync(content, cultures, publishedRouter, umbracoContext, contentService, textService, variationContextAccessor, logger, uriUtility, publishedUrlProvider))
```
### My contention...
My contention is that for a non-variant site, this presents a lot of unnecessary and confusing information for editors, which promotes blindness to when there is an actual routing problem!

Also the information is incorrect

![image](https://github.com/umbraco/Umbraco-CMS/assets/260802/652e96ec-f097-46b7-b590-d5245fb8fdbb)

Here it thinks the node would collide with /Home/Destinations page... but it won't... because the first node in Umbraco is routed as / and not as it's Url Name, but also because the url starts with /us that will be route via culture and hostnames to the US site, so it is untrue to say this will clash!

## Bit that needs thought
The assumption which I think needs a bit of thought is this:

```
 // we don't know if the branch to this content is invariant, so we need to ask
        // for URLs for all cultures.
```
I think firstly for both variant and non-variant sites - you probably only want to do this for cultures that have been assigned to nodes in the branch of the site, rather than 'all languages in Umbraco'. (you can see in the screenshot I added 'en' language and it immediately appeared as a Url, even though it's not setup to be routed anywhere...)

There is a database table called umbracoDomain that stores the mapping of languageId to NodeId...

![image](https://github.com/umbraco/Umbraco-CMS/assets/260802/63b6d669-7764-4ce9-9631-a8d575330ef4)

(it also seem to have entries for the nodes *48036 for the context of the backoffice)

I wonder if it could be used, along with the Path property of the Umbraco item in the backoffice, eg -1, 1234, 134566 - to lookup the cultures available to be varied upon on the current branch rather than using all languages?

## good default behaviour for non variant sites?
But if this is actually the desired display for Variant-based sites... could we do something different for non-variant sites?

Could we make it configurable or perhaps only triggered if there is a Document Type on the site set to 'vary by culture' ?

## Workaround for future non-variant Umbracians coming here with a similar fear of these extraneous Url messages

The way I'm working around this, for anyone coming to this issue with the same problem is to handle the EditorSendingContentNotification and get the parentNode and use the DomainUtility GetCultureFromDomains() to find the current branches culture, and then remove all the other entries for other cultures... but this, I think will only work if a single culture is assigned to a branch... but if it's useful it looks something like this:

```
public class EditorSendingContentNotificationHandler : INotificationHandler<SendingContentNotification>
{
	private readonly IUmbracoContextAccessor _umbracoContextAccessor;

	public EditorSendingContentNotificationHandler(IUmbracoContextAccessor umbracoContextAccessor)
	{
		_umbracoContextAccessor = umbracoContextAccessor;
	}
	public void Handle(SendingContentNotification notification)
	{
		if (notification != null && notification.Content != null)
		{
			// get culture of parent?
			string? branchCulture = "";
			if (_umbracoContextAccessor.TryGetUmbracoContext(out IUmbracoContext umbracoContext))
			{
				var parentNode = umbracoContext?.Content?.GetById(notification.Content.ParentId ?? 0);
				if (parentNode != null)
				{
					branchCulture = parentNode.GetCultureFromDomains();
				}
			}
			if (!string.IsNullOrEmpty(branchCulture) && notification.Content.Urls != null)
			{
				List<UrlInfo> filteredUrlInfos = new List<UrlInfo>();
				foreach (var url in notification.Content.Urls)
				{
					if (url.Culture == branchCulture)
					{
						filteredUrlInfos.Add(url);
					}
				}

				notification.Content.Urls = filteredUrlInfos.ToArray();
			}
			if (notification.Content.ContentTypeAlias == "destinationRegion")
			{
				var messages = new List<UrlInfo>();
				messages.Add(UrlInfo.Message("Destination Region pages have no URL."));
				notification.Content.Urls = messages.ToArray();
			}

		}

	}
}
```
Et Voila! the editor sees one Url that is relevant for the territory they are on, as you'd expect.

### Steps to reproduce

Create a New Umbraco Site with a single domain pointing at it
Create Four homepages at the root of Umbraco and some pages underneath
DO NOT make any of the document types vary by culture - in fact don't even whisper variants to anyone when you are creating them, don't think about variants, just hum beatles songs or something.
Add Four languages to the Umbraco Site
Assign one of each of those languages to one of each of the four homepages, and use one-level paths for each language site other than the first to define the difference
![image](https://github.com/umbraco/Umbraco-CMS/assets/260802/8f4c2421-d7c5-4a27-813d-ffdd2980c2b8)

Now visit a page underneath any of the sites, switch to the Info Content App, and examine the links...
... see how you have a link for every language in the site!

Is it confusing?

Add more languages to Umbraco - see how they immediately appear as links,even though they are not assigned to domains.

**Comment by Zeegaan** [comment]
Wierd one, I am seeing to links always (the default language, and then the given culture i assigned via domain)
But the other links aren't there for me 😅 

![image](https://github.com/umbraco/Umbraco-CMS/assets/70372949/0b51be6d-0574-4ec0-8441-26f6790fc862)

**Comment by marcemarc** [comment]
@Zeegaan so for your TwoLanguage site, and your ChildTo page - why are you seeing
'This document is published but is not in the cache'
for the ChildTwo page
for da / sv/ am languages?

![image](https://github.com/umbraco/Umbraco-CMS/assets/260802/a43f2334-9648-4d0e-99ef-af8f1d3d3a3d)

Those cultures arent' added to the TwoLanguage page, and your Umbraco site isn't using Variants... so that is what I'm reporting it's really confusing for the editor that only edits the 'Two language' branch, to have these additional messages about cultures! that sound like something is wrong... but it is not!

regards

Marc

**Comment by Zeegaan** [comment]
Aha good points, will take this up with the team on what to do about this 😁  💪

**Comment by marcemarc** [comment]
thanks @Zeegaan !! - I think it might have been like this since V8, it's only now that we are migrating all these V7 sites that are coming across these anomalies for non variant sites

But appreciate you taking a look!

**Comment by bjarnef** [comment]
I noticed something similar in a multi-site solutions where we have 9 languages, but use the old language structure as each site needed its own site structure.

But on each node we then had 8 empty links 😅

The workaround we have for now.

```
public class EditorSendingContentNotificationHandler : INotificationHandler<SendingContentNotification>
{
    private readonly IPublishedSnapshotAccessor _publishedSnapshotAccessor;

    public EditorSendingContentNotificationHandler(
        IPublishedSnapshotAccessor publishedSnapshotAccessor)
    {
        _publishedSnapshotAccessor = publishedSnapshotAccessor;
    }

    public void Handle(SendingContentNotification notification)
    {
        if (notification.Content.ContentTypeAlias.Equals(EmployeeArchive.ModelTypeAlias) ||
            notification.Content.ContentTypeAlias.Equals(Employee.ModelTypeAlias) ||
            notification.Content.ContentTypeAlias.Equals(LocationArchive.ModelTypeAlias) ||
            notification.Content.ContentTypeAlias.Equals(Location.ModelTypeAlias) ||
            notification.Content.ContentTypeAlias.Equals(TestimonialArchive.ModelTypeAlias) ||
            notification.Content.ContentTypeAlias.Equals(Testimonial.ModelTypeAlias) ||
            notification.Content.ContentTypeAlias.Equals(TopicsArchive.ModelTypeAlias) ||
            notification.Content.ContentTypeAlias.Equals(Topic.ModelTypeAlias))
        {
            notification.Content.Urls = null;
        }
        else
        {
            notification.Content.Urls = notification.Content.Urls?.Where(x => x.IsUrl).ToArray();
        }
    }
}
```

**Comment by marcemarc** [comment]
@bjarnef Thanks, but I think that will hide all the non url messages, which might be genuine warnings!!! here we want messages that aren't necessarily Urls, just not for the cultures that don't apply - if you look at my workaround above I use GetCultureFromDomains on the parent Item, so that only the language specific version is shown and it can be either a message or a url!

**Comment by bjarnef** [comment]
@marcemarc yes, I think this was a quick workaround and better UI/UX to the customer.
It does show the "_This item is not published_" message, but I haven't checked the "_This document is published but is not in the cache_" or other error related message.

I always find it a bit odd to show error message inline here. Could probably be redesigned with a warning/error icon instead opening a details overlay with more error/warning details or an inline warning/error notification, similar to languages when changing default language?

![image](https://github.com/umbraco/Umbraco-CMS/assets/2919859/13a25b6c-09a5-456d-a3a0-37e30aeffa47)

**Comment by iOvergaard** [comment]
Hi @marcemarc,

Revisiting this one — the specific behaviour you and others reported here (every configured language showing up as a link or "not in the cache"/"not created" message on non-variant content, regardless of domain assignment) no longer happens on current versions.

The backoffice's Info tab now goes through a different code path than the one you dug into back then, and as of v17.2.0 it filters cultures for invariant content down to just the one(s) actually assigned via domains/hostnames to that branch (falling back to the default language if none are assigned) — matching what you originally asked for. Unroutable results for those aren't reported as messages either, since for invariant content that's just "no domain for this culture," not a routing problem.

Worth noting this actually regressed briefly during a V15 rework and was re-reported as [#21459](https://github.com/umbraco/Umbraco-CMS/issues/21459), fixed properly in [#21473](https://github.com/umbraco/Umbraco-CMS/pull/21473). So it should be solid from v17.2.0 onward and in the current v18 line.

Closing this out, but please reopen if you still see the old behaviour on a current version.

**Comment by marcemarc** [comment]
Thanks @iOvergaard , one of the last sites we have left to upgrade to v17, is the one that sparked the original issue and deep dive, but from you report above, sounds like it will be ok!

**Comment by bjarnef** [comment]
@iOvergaard in v17.5.0 it seems to work okay as long a hostname is assigned.

However when customizing URL provider to return an external/absolute URL or replacing hostname, I get "not in the cache"/"not created" messages although the `UrlInfo` returned seems to be correct.

https://github.com/umbraco/Umbraco-CMS/issues/23028#issuecomment-4909426912

**Comment by iOvergaard** [comment]
@bjarnef, the gist of the Links box is that it works like this:

> as of v17.2.0 it filters cultures for invariant content down to just the one(s) actually assigned via domains/hostnames to that branch (falling back to the default language if none are assigned)

What I gather is that, by design, you must assign a hostname. I am not sure whether custom URL providers ought to work in this instance. I hope @kjac gets back to you over on https://github.com/umbraco/Umbraco-CMS/issues/23028#issuecomment-4909426912 to clarify further.

**Comment by bjarnef** [comment]
@iOvergaard yeah, it just has `/` assigned as hostname, and works fine with default URL provider or the random guid as relative URL as @kjac showed. Not sure if there are some issues, when I use absolute URLs.

Two use cases:

- Replace hostname - same URL as default URL provider, just a different than requested for backoffice (mainly because of headless frontend)
- External URL - for specific content type use a provided value in `externalUrl` property if parsed correct as an absolute URL.
